### ISSUE TITLE
change coq-prog-name from _CoqProject

### ISSUE THREAD

**Issue by SimonBoulier** [ProofGeneral/PG#60]
Hi,
Is it possible to change `coq-prog-name` from the _CoqProject?
I tried to add:
  `COQTOP = ~/.opam/coq/bin/coqtop`
  `COQC = ~/.opam/coq/bin/coqc`
to my _CoqProject but it dosn't work (PG still uses the coqtop of my system).
Thank you very much!

**Comment by cpitclaudel** [comment]
Hi Simon,

Thanks for the suggestion! Is this syntax documented somewhere?

Clément.

**Comment by SimonBoulier** [comment]
Hum... no I don't think so.
Je prenais juste mes désirs pour la réalité !

**Comment by erikmd** [comment]
Hi,

The `_CoqProject` syntax appears to be documented in
https://coq.inria.fr/distrib/V8.5/refman/Reference-Manual017.html#Makefile
and in the output of the command `coq_makefile --help`

Indeed these `_CoqProject` environment variables are not supported by PG currently.
Still, you can use one of the 2 solutions below:

1) Write a file `.dir-locals.el` in the root folder of your Coq project with contents
```elisp
((coq-mode . ((coq-prog-name "~/.opam/coq/bin/coqtop"))))
```
then restart Emacs (or run
`M-x proof-shell-exit`
`M-x normal-mode`
in the Coq buffer before restarting the Coq process).

2) or if you want to automatically set the variable `coq-prog-name` at
Emacs startup according to the currently-enabled OPAM switch, you can
add in your `.emacs.d/init.el` a piece of code such as:
```elisp
(with-temp-buffer (insert-file-contents "~/.opam/config")
                  (string-match "^switch: \\\"\\\([^\\\"]+\\\)\\\""
                                (buffer-string))
                  (setq coq-prog-name
                        (concat "~/.opam/"
                                (match-string 1 (buffer-string))
                                "/bin/coqtop")))
```
Hoping that this helps,

Erik

**Comment by JasonGross** [comment]
@cpitclaudel , it's documented in `coq_makefile --help`:
```
[VARIABLE = value]: Add the variable definition "VARIABLE=value"
```

**Comment by SimonBoulier** [comment]
@erikmd, Thank you for your solutions. 1) is cool.
However, it didn't work on the first attempt so I used the version from the HoTT library:
`((coq-mode . ((eval . (let ((default-directory (locate-dominating-file
						buffer-file-name ".dir-locals.el")))
			(make-local-variable 'coq-prog-name)
			(setq coq-prog-name (expand-file-name "~/.opam/coq/bin/coqtop")))))))`
even if I don't understand all what it does.

I didn’t tried 2) because I also need to switch between `coqtop` and `hoqtop` from the HoTT library (yes, a bit painful...).


Whatever, would be a very nice feature for PG!

**Comment by cpitclaudel** [comment]
@Jason, thanks for the pointer. Life would be simpler if we didn't need to reimplement a _CoqProject parser. Maybe I'll find time to extend the current implementation to handle these variables.

@erikmd, thanks for your help!

@SimonBoulier, this looks needlessly complicated, actually :) That bit of code is exactly equivalent to the one that Erik sent. The extra complications are there because hott uses a relative file name, not an absolute one. In your case, though, Erik's code should be enough.

Try using Erik's code again, saving the `.dir-locals.el`, and finally running `M-x revert-buffer` on your Coq source file.

**Comment by SimonBoulier** [comment]
@cpitclaudel  The error is:
`apply: Wrong type argument: stringp, ("~/.opam/coq/bin/coqtop")`

**Comment by erikmd** [comment]
Hi,

Which version of Emacs do you have? Mine is GNU Emacs 24.3.1, and the
contents I had proposed for .dir-locals.el was working for me...

However, could you try the following variant?

((coq-mode . ((coq-prog-name . "~/.opam/coq/bin/coqtop"))))

where these is a dot between the variable name and the value.

Regards,

Erik

Le jeudi 25 février 2016 à 06:08 -0800, SimonBoulier a écrit :
> @cpitclaudel  The error is:
> `apply: Wrong type argument: stringp, ("~/.opam/coq/bin/coqtop")`

-- 
Érik Martin-Dorel
erik@martin-dorel.org
http://erik.martin-dorel.org/

**Comment by SimonBoulier** [comment]
It works with the dot!
I have Emacs 24.4.1.
Thank you very much.

Simon

**Comment by JasonGross** [comment]
@cpitclaudel , useful variables to read from the _CoqProject file include `COQC`, `COQTOP`, `COQBIN`, and possibly `COQPATH`.

**Comment by Matafou** [comment]
Hi,
I am pretty sure that coq-prog-name (and COQXXX variables) should not be
set in .CoqProject file. This project file is supposed to work for all
users and is typically supposed to be included in the distribution of a
library (neither included in its repository). On the contrary coq-prog-name
is set differently in each user configuration and its configuration should
not appear in git or distributions.

With coqide the problem does not occur since the coqide executables that
know where to find its corresponding coqtop. This is a problem that only
arises with external IDEs.

The .dir-locals.el can be a solution, but developpers may also want to use
this file in their archives (to force some emacs config for instance). In
this case it is not a good idea to put local configuration in it. Maybe we
should add a special .pg.el file?
P.

Le mar. 8 mars 2016 à 20:18, Jason Gross <notifications@github.com> a
écrit :

**Comment by JasonGross** [comment]
@Matafou This is not the case for projects like HoTT, where `coq-prog-name` should be set to `./hoqtop`, which picks up the system Coq in the `./configure` step.

**Comment by cpitclaudel** [comment]
I talked to Matthieu about this at the DeepSpec kickoff meeting. The format of _Coqproject is rather hard to parse ATM; it would be much simpler for coqdep to give us a structured version of it, say in JSON or XML. Is HoTT planning to update to 8.6 when it comes out? If so, we could try to get this in 8.6.

**Comment by aspiwack** [comment]
> it would be much simpler for coqdep to give us

I'm curious: how did coqdep enter this conversation?

> JSON or XML.

Teeth gnashing words :-p . If we are speaking of a project file which is supposed to be edited by a human please avoid Json at all cost (yes, that's experience speaking…). Xml is not perfect for that either, but it's ok (at least it has comments!). Structured formats which are usable for human and enjoy nice support include Toml and Yaml.

If it's just a serialisation format, then ok, I guess (though S-expr are better :-p ).

**Comment by cpitclaudel** [comment]
> I'm curious: how did coqdep enter this conversation?
My bad. I meant coq_makefile.

> If it's just a serialisation format, then ok, I guess (though S-expr are better :-p ).
I wasn't thinking of changing the format; just of getting an easy way for us to parse it :) And I'm all for s-exprs :)

**Comment by erikmd** [comment]
Hi, I would just like to give a +1 for @Matafou 's suggestion:
> The `.dir-locals.el` can be a solution, but developpers may also want to use this file in their archives (to force some emacs config for instance). In this case it is not a good idea to put local configuration in it. Maybe we should add a special `.pg.el` file?

For simplicity, this `.pg.el` file could have the same (S-expr) syntax as `.dir-locals.el` and thereby contain for HoTT:
```
((coq-mode . ((coq-prog-name . "./hoqtop"))))
```
BTW I've just noticed that a similar feature (for an extra Directory Local Variables file `.dir-locals-2.el`) has been added in Emacs' master branch:
https://github.com/emacs-mirror/emacs/commit/da976cff352bdea6adc2667582a56eb4061bb5f6#diff-725cb3c0ff9b6d12e3ace616eb5b0d0dR3723
and the documentation says:
> This essentially a second file that can be used like `dir-locals-file`, so that users can have specify their personal dir-local variables even if the current directory already has a `dir-locals-file` that is shared with other users (such as in a git repository)

So maybe a `.pg.el` file would be the proper way to address #60 ?

**Comment by Matafou** [comment]
Or maybe we should document that the right way to set coq-prog-name locally
(and any other option that should not be versioned) is to use a
".dir-locals-2.el"?

But I guess we want to set coq-prog-name for other ides as well (except
coqide). So maybe lisp syntax is not appropriate here and something more
agnostic would be better.

P.



2016-06-14 22:19 GMT+02:00 Erik Martin-Dorel <notifications@github.com>:

**Comment by erikmd** [comment]
Hi @Matafou
> Or maybe we should document that the right way to set coq-prog-name locally (and any other option that should not be versioned) is to use a ".dir-locals-2.el"?

Yes maybe. However I'm afraid that native support for `.dir-locals-2.el` won't be available in Emacs before some time, as it is not available in Emacs 24, and this change was not yet released in any version of Emacs 25, according to https://github.com/emacs-mirror/emacs/commit/71b20b31c8ee25bcd0de1a954c9ca7b49140b3b2

So: Either we keep `.dir-locals-2.el` as a filename and (re)implement this kind of support in PG (plus some code *à la* `(fboundp 'dir-locals--all-files)` or `(boundp 'dir-locals-file-2)` to test if this feature is not already provided by Emacs, to cope with future versions of Emacs). Or we choose another filename, such as `.pg.el`. I don't know what would be the best choice(?)

> But I guess we want to set coq-prog-name for other ides as well (except coqide). So maybe lisp syntax is not appropriate here and something more agnostic would be better.

Indeed... But on the other hand, the config file we are talking about is not intended to be included in a Git repository or distributed among the sources of a Coq library, as it specifically deals with user-defined config... So maybe sticking to emacs-lisp format is not an issue. But of course, using an elisp config file (rather than XML or so) will require the usual check regarding safe values.

**Comment by cpitclaudel** [comment]
Hmm. I not sure I see the point of the `.pg.el` (or .dir-locals-2.el`). What was wrong with `.dir-locals.el`?

**Comment by cpitclaudel** [comment]
Oh, sorry, I should have read more closely. I really don't think we should set up a complex mechanism for this to work around a relatively minor annoyance. In most Coq repos that I'm familiar with the code isn't at the top-level; it's in a subdirectory instead (`src/` or such). Thus `.dir-locals.el` files can/should be moved to the `src/` directory, and then user customizations can be put at the root.

Alternatively, it's always possible to create an extra folder and move the provect tree one level deeper, leaving space for a user-only `.dir-locals.el`.

**Comment by erikmd** [comment]
Hi Clément, I agree with your comment.
Anyway, as Pierre suggested, I believe we should document somewhere (e.g. in the FAQ or in the user manual) that the recommended way to configure `coq-prog-name` is by using a `.dir-locals.el` file... and afterwards we'll be able to close this issue

**Comment by Matafou** [comment]
Agreed.

2016-07-16 21:39 GMT+02:00 Erik Martin-Dorel <notifications@github.com>:

**Comment by erikmd** [comment]
OK, so I've added a comment in both the FAQ and the user manual.
Closed by 026a3a7f984259c65ffa707c6bb8196c311a2f4b.
